kroxylicious/src/main/java/io/kroxylicious/proxy/internal/net/EndpointRegistry.java

```diff
@@ -0,0 +1,417 @@
+        if (vcr == null) {
+            // cluster not currently registered
+            deregisterFuture.complete(null);
+            return deregisterFuture;
+        }
+
+        var updated = vcr.deregistrationStage().compareAndSet(null, deregisterFuture);
+        if (!updated) {
+            // cluster de-registration already in progress.
+            return vcr.deregistrationStage().get();
+        }
+
+        var unused = vcr.registrationStage()
+                .thenCompose((u) -> deregisterBinding(virtualCluster, binding -> binding.virtualCluster().equals(virtualCluster))
+                        .handle((unused1, t) -> {
+                            registeredVirtualClusters.remove(virtualCluster);
+                            if (t != null) {
+                                deregisterFuture.completeExceptionally(t);
+                            }
+                            else {
+                                deregisterFuture.complete(null);
+                            }
+                            return null;
+                        }));
+        return deregisterFuture;
+    }
+
+    /* test */ boolean isRegistered(VirtualCluster virtualCluster) {
+        return registeredVirtualClusters.containsKey(virtualCluster);
+    }
+
+    /* test */ int listeningChannelCount() {
+        return listeningChannels.size();
+    }
+
+    private CompletionStage<Endpoint> registerBinding(Endpoint key, String host, VirtualClusterBinding virtualClusterBinding) {
+        Objects.requireNonNull(key, "key cannot be null");
+        Objects.requireNonNull(virtualClusterBinding, "virtualClusterBinding cannot be null");
+        var virtualCluster = virtualClusterBinding.virtualCluster();
+
+        var lcr = listeningChannels.computeIfAbsent(key, k -> {
+            // the listening channel doesn't exist, atomically create a record to represent it and request its binding to the network.
+            var future = new CompletableFuture<Channel>();
+            var r = ListeningChannelRecord.create(future.exceptionally(t -> {
+                // Handles the case where the network bind fails
+                listeningChannels.remove(key);
+                if (t instanceof RuntimeException re) {
+                    throw re;
+                }
+                else {
+                    throw new RuntimeException(t);
+                }
+            }));
+
+            queue.add(new NetworkBindRequest(key.bindingAddress(), key.port(), virtualCluster.isUseTls(), future));
+            return r;
+        });
+
+        if (lcr.unbindingStage().get() != null) {
+            // Listening channel already being unbound, chain this request to the unbind stage so it completes later.
+            return lcr.unbindingStage().get().thenCompose(u -> registerBinding(key, host, virtualClusterBinding));
+        }
+
+        return lcr.bindingStage().thenApply(c -> {
+            synchronized (lcr) {
+                var bindings = c.attr(CHANNEL_BINDINGS);
```

Comment:
We puzzled over this for a while. Maybe renaming `c` to something like `acceptorChannel` would help make it clear we store the map on a particular channel.

Is the purpose of this map to enable us to tear down a listener when the last binding is deregistered from it? (one acceptor may be routing to multiple bindings using SNI)

Reply:
renamed parameters so and added comment.